velox/functions/sparksql/SplitFunctions.cpp

```diff
@@ -16,124 +16,195 @@
+  // Split with a non-empty pattern.
+  void splitInner(
+      StringView input,
+      const StringView delim,
+      int64_t limit,
+      vector_size_t row,
+      exec::VectorWriter<Array<Varchar>>& resultWriter) const {
+    // Add new array (for the new row) to our array vector.
+    resultWriter.setOffset(row);
+    auto& arrayWriter = resultWriter.current();
+
+    // Trivial case of converting string to array with 1 element.
+    if (limit == 1) {
+      arrayWriter.add_item().setNoCopy(input);
+      resultWriter.commit();
+      return;
+    }
+
+    // We walk through our input cutting off the pieces using the delimiter and
+    // adding them to the elements vector, until we reached the end of the
+    // string or the limit.
+    int32_t addedElements{0};
+    bool emptyDelim = delim.size() == 0 ? true : false;
+    auto* re = cache_.findOrCompile(delim);
+    const auto re2String = re2::StringPiece(input.data(), input.size());
+    size_t pos = 0;
+    const char* start = input.data();
+    re2::StringPiece subMatches[1];
+    while (re->Match(
+        re2String, pos, input.size(), RE2::Anchor::UNANCHORED, subMatches, 1)) {
+      const auto fullMatch = subMatches[0];
+      auto offset = fullMatch.data() - start;
+      const auto size = fullMatch.size();
+
+      if (offset >= input.size()) {
+        break;
+      }
+
+      arrayWriter.add_item().setNoCopy(
+          StringView(input.data() + pos, offset - pos));
+      pos = offset + size;
+      ++addedElements;
+      // If the next element should be the last, leave the loop.
+      if (addedElements + 1 == limit) {
+        break;
+      }
+    }
+
+    // Add the rest of the string and we are done.
+    // Note, that the rest of the string can be empty - we still add it.
+    arrayWriter.add_item().setNoCopy(
+        StringView(input.data() + pos, input.size() - pos));
+    resultWriter.commit();
   }
+
+  mutable functions::detail::ReCache cache_;
 };
 
-/// The function returns specialized version of split based on the constant
-/// inputs.
-/// \param inputArgs the inputs types (VARCHAR, VARCHAR, int64) and constant
-///     values (if provided).
 std::shared_ptr<exec::VectorFunction> createSplit(
     const std::string& /*name*/,
     const std::vector<exec::VectorFunctionArg>& inputArgs,
     const core::QueryConfig& /*config*/) {
-  BaseVector* constantPattern = inputArgs[1].constantValue.get();
-
-  if (inputArgs.size() > 3 || inputArgs[0].type->isVarchar() ||
-      inputArgs[1].type->isVarchar() || (constantPattern == nullptr)) {
-    return std::make_shared<Split>();
-  }
-  auto pattern = constantPattern->as<ConstantVector<StringView>>()->valueAt(0);
-  if (pattern.size() != 1) {
-    return std::make_shared<Split>();
-  }
-  char charPattern = pattern.data()[0];
-  // TODO: Add support for zero-length pattern, 2-character pattern
-  // TODO: add support for general regex pattern using R2
-  return std::make_shared<SplitCharacter>(charPattern);
```

Comment:
Please add some checks for the input argument size and types.
https://github.com/facebookincubator/velox/pull/8825/files#diff-ebc054da2ecc11ac3cf20d47cbc112f9c3b9cf3d5f1784b2149fb0f4e997a2d8R204-R216